sdk/clientcore/core/src/main/java/io/clientcore/core/utils/CoreUtils.java

```diff
@@ -498,6 +500,33 @@ public static <T> T decodeNetworkResponse(BinaryData data, ObjectSerializer seri
         }
     }
 
+    /**
+     * Instantiates an {@link HttpResponseException} for unexpected responses.
+     *
+     * @param responseCode The HTTP response code.
+     * @param response The HTTP response.
+     * @param data The body of the HTTP response.
+     * @param decodedValue The decoded value.
+     * @return An {@link HttpResponseException} for unexpected responses.
+     */
+    public static HttpResponseException instantiateUnexpectedException(int responseCode, Response<BinaryData> response,
+        BinaryData data, Object decodedValue) {
+        StringBuilder exceptionMessage = new StringBuilder("Status code ").append(responseCode).append(", ");
+        String contentType = response.getHeaders().getValue(HttpHeaderName.CONTENT_TYPE);
+        if ("application/octet-stream".equalsIgnoreCase(contentType)) {
+            String contentLength = response.getHeaders().getValue(HttpHeaderName.CONTENT_LENGTH);
+            exceptionMessage.append("(").append(contentLength).append("-byte body)");
+        } else if (data == null || data.toBytes().length == 0) {
+            exceptionMessage.append("(empty body)");
+        } else {
+            exceptionMessage.append('"').append(new String(data.toBytes(), StandardCharsets.UTF_8)).append('"');
+        }
+        if (decodedValue instanceof IOException || decodedValue instanceof IllegalStateException) {
```

Comment:
I don't think this logic is correct with what has changed between annotation processor and RestProxy. The decoded value should be the declared exception type, ex `ErrorValue`.